Support unicode emojis and remove emojify.js

mrsdizzie opened the pull request:
This PR replaces all use of emojify.js and adds unicode emoji support to various areas of gitea.

This works in a few ways:

First it adds emoji parsing support into gitea itself. This allows us to

 * Render emojis from valid alias (```:smile:```)
 * Detect unicode emojis and let us put them in their own class with proper aria-labels and styling
 * Easily allow for custom "emoji"
 * Support all emoji rendering and features without javascript
 * Uses plain unicode and lets the system render in appropriate emoji font
 * Doesn't leave us relying on external sources for updates/fixes/features

That same list of emoji is also used to create a json file which replaces the part of emojify.js that populates the emoji search tribute. 

For custom "emoji" it uses a pretty simple scheme of just looking for /emojis/img/name.png where name is something a user has put in the "allowed reactions" setting we already have. The gitea reaction that was previously hard coded into a forked copy of emojify.js is included and works as a custom reaction under this method.

The emoji data sourced here is from https://github.com/github/gemoji which is the gem library Github uses for their emoji rendering (and a data source for other sites). So we should be able to easily render any emoji and :alias: that Github can, removing any errors from migrated content. They also update it as well, so we can sync when there are new unicode emoji lists released.

I've included a slimmed down and slightly modified forked copy of https://github.com/knq/emoji to make up our own emoji module. The code is pretty straight forward and again allows us to have a lot of flexibility in what happens.

I had seen a few comments about performance in some of the other threads if we render this ourselves, but there doesn't seem to be any issue here. In a test it can parse, convert, and render 1,000 emojis inside of a large markdown table in about 100ms on my laptop (which is many more emojis than will ever be in any normal issue). This also prevents any flickering and other weirdness from using javascript to render some things while using go for others.

Includes the optional download of a fallback emoji webfont for small number of users who don't already have one installed.

Fixes: https://github.com/go-gitea/gitea/issues/9182
Fixes: https://github.com/go-gitea/gitea/issues/8974
Fixes: https://github.com/go-gitea/gitea/issues/8953
Fixes: https://github.com/go-gitea/gitea/issues/6628
Fixes: https://github.com/go-gitea/gitea/issues/5130

silverwind commented:
This sounds like a clearly better alternative then https://github.com/go-gitea/gitea/pull/11031. Maybe you can port the `Tribute` extraction from my PR to yours. Your data source seems fine and more maintained than mine, albeit it's a bit smaller (35kB vs 40kB).

mrsdizzie commented:
I wanted to update our copy of tribute and was actually going to ask you to help with that after this -- I noticed some bugs with ours (like you can't delete) that seem to be fixed in newer versions. I'll look at how your PR does that (I don't like sticking all of that in footer template anyway)

silverwind commented:
Suggestions:

- extract tribute to its own file in `web_src/features/tribute` similar to my PR.
- `import map from './path/to/json'` to load the data directly into the webpack bundle instead of ajax.
- maybe use my version of the tribute `values` function:

````js
    values(query, cb) {
      const matches = [];
      for (const name of allNames) {
        if (name.startsWith(query)) {
          matches.push(name);
          if (matches.length > 5) break;
        }
      }
      cb(matches);
    },
````

Difference here is `.startsWith` which I find more useful than `indexOf` which can match too much on short strings.

mrsdizzie commented:
Thank you I am not familiar with javascript at all! I've added you as a collaborator if you want to push those changes, if not I can try to get to them in a bit

silverwind commented:
I'll have a look at extracting tribute tomorrow. I guess I would also move `emoji.json` to a better location, maybe the `assets` directory. [My data source](https://unpkg.com/emojilib@2.4.0/simplemap.json) for reference.

6543 commented:
looks like emoji.json exist twice ? `public/emoji.json` and `public/emoji/emoji.json`

lunny commented on build/generate-emoji.go:
```diff
@@ -0,0 +1,134 @@
+)
+
+// Gemoji is a set of emoji data.
+type Gemoji []Emoji
```
Why not import `modules/emoji`?

mrsdizzie replied:
modules/emoji loads the Gemoji  []struct that this file creates and so if you want to make a change here (like add a field to the struct) it will give an error about wrong number of fields in the existing data and you would need to manually edit it. 

Since this is just a generate script and not run often it seemed easier to just redefine the struct here and avoid that problem of requiring the module that it tries to modify.

silverwind commented:
Rebased to fix issue references in orginal commit message.

lafriks commented:
I think it's missing aliases for `:)`, `;)` etc

silverwind commented:
@lafriks those were previously [disabled](https://github.com/go-gitea/gitea/blob/743022116df236c0a77c020ad142a6e21b394320/web_src/js/index.js#L2473) corresponding to [this](https://github.com/joypixels/emojify.js/blob/58e651c6e9b89cde35e6ccfa4f1e43374982789c/src/emojify.js#L67).

mrsdizzie commented:
> I think it's missing aliases for `:)`, `;)` etc

Those are emoticons which are technically different and often people want to use them as is as a plain text alternative to an emoji. I purposely didn't  add support for those for that reason. 

If you wanted to add a smile you can start typing ```:smile``` or insert the emoji normally

6543 commented on build/generate-emoji.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+const hdr = `
+package emoji
```
Isnt the copiright header first and the package definde?

mrsdizzie replied:
Fixed

mrsdizzie commented on web_src/less/_base.less:
```diff
@@ -1231,10 +1231,10 @@ i.icon.centerlock {
-    background-size: contain;
-    font-size: 1.5em !important;
-    vertical-align: middle !important;
+    font-size: 1.2em;
```
@silverwind I liked this size better at 1.5em

silverwind replied:
I copied that size from GitHub's definition, but feel free to revert. I did not test this yet. My thought is that if a page contains regular unicode emoji that are not rendered from tokens, those would not have the `.emoji` class and would render with a different size, but correct me if I'm wrong and the backend is able to detect those as well.

silverwind replied:
GitHub seems to be able to detect unicode emoji and wrap them in a `<span>`:

From `:camel`: 🐫 
Pasted: 🐫

mrsdizzie replied:
I did make it able to detect all emoji and apply the class. I can't seem to make a working link to a line in a PR but it is emojiRegex inside modules/markup/html.go and emojiProcessor for that same file. Also I used this to test it https://regex101.com/r/XO3FWR/2

I like adjusting the size because its harder to see otherwise (many other sites adjust it from 1.2-1.5em as you've seen Github uses 1.2em I think gitlab uses 1.5em).

silverwind replied:
If you already wrap existing unicode-emojis in markdown, we should be fine.

I'll test more later to form an opinion on the font size.

guillep2k replied:
Some patterns are not matched, or matched only partially (where adding `<span>` would break the composition semantings). Is this intentional, a know limitation, ...?

![image](https://user-images.githubusercontent.com/18600385/79034621-c42b5200-7b8d-11ea-8cf7-3300bc36697b.png)

Note: I've followed the test URL https://regex101.com/r/XO3FWR/2 as posted above.

mrsdizzie commented:
> looks like emoji.json exist twice ? `public/emoji.json` and `public/emoji/emoji.json`

fixed

silverwind commented:
So this 394 kB (ungzipped) seems a bit on the heavy side. Would you mind building a lightweight name to unicode mapping format for the frontent (pretty much like https://unpkg.com/emojilib@2.4.0/simplemap.json)?

mrsdizzie commented:
> So this 394 kB (ungzipped) seems a bit on the heavy side for frontend inclusion. Would you mind building a lightweight name to unicode mapping format for the frontent (pretty much like https://unpkg.com/emojilib@2.4.0/simplemap.json)?

Sure I'll make something like that that, but for my own knowledge: It is 35kb gzipped, which I think anybody who cares about kb of size should enabled  -- unless I'm misunderstanding and the size issue is related to something else about performance? Does it have some noticeable effect after it is loaded? I'm just not familiar with that area at all.

silverwind commented:
Yeah, it gzips well, but the reduced version is 40kb without and 12kB with gzip.

It's not a big deal but I'd still prefer the smaller version because every byte matters, especially for slow connections. With a active service worker (the default), the content should only be downloaded once every 24h at most, so the price is not paid every page load for the common case.

Do you forsee that we will need the additional data on the frontend? If so, keep it.

mrsdizzie commented:
> Yeah, it gzips well, but the reduced version is 40kb without and 12kB with gzip.
> 
> It's not a big deal but I'd still prefer the smaller version because every byte matters, especially for slow connections. With a active service worker (the default), the content should only be downloaded once every 24h at most, so the price is not paid every page load for the common case.
> 
> Do you forsee that we will need the additional data on the frontend? If so, keep it.

In thinking about it I don't see us actually doing anything with most of the fields in general (category, tags, unicode and ios version, skin_tones bool). So good suggestion I've just removed those.

It is now 27kb gzipped. It would be 18kb without the aliases but I'd rather have them to search and keep the json and what go uses in sync than save the additional 9kb.

Since it loads after the page has loaded it shouldn't visually effect or slow rendering on a slow connection either (I tried on the chrome 3g simulator and says it takes 750ms to load in the background after the page is rendered).

6543 commented on templates/repo/issue/view_content/add_reaction.tmpl:
```diff
@@ -4,16 +4,16 @@
 		{{range $value := AllowedReactions}}
 			{{if eq $value "hooray"}}
-				<div class="item" data-content="hooray">:tada:</div>
+				<div class="item reaction" data-content="hooray">{{ReactionToEmoji "tada"}}</div>
```
If we support alias now the whole if statement is pointles and can be removed

6543 replied:
can we migrate them ?

-> I think this is worth it's own pull

mrsdizzie replied:
fixed -- add custom aliases before we generate the files

6543 commented:
@mrsdizzie we realy need a fallback:

![Screenshot at 2020-04-11 02-44-04](https://user-images.githubusercontent.com/24977596/79031346-66ebcc80-7b9e-11ea-8b18-2f2d4433bfc6.png)

mrsdizzie commented:
You don't have a good emoji font on your system?

6543 commented:
> You don't have a good emoji font on your system?

I dont think we should require that

6543 commented:
custom emoji not renderd on review comment:

![Screenshot at 2020-04-11 03-04-24](https://user-images.githubusercontent.com/24977596/79031736-3bb6ac80-7ba1-11ea-9e7d-8a06add2174b.png)

silverwind commented:
@6543 try installing Noto Emoji. Recent versions of Ubuntu come with it by default. What's your distro?

mrsdizzie commented on modules/markup/html.go:
```diff
@@ -385,6 +405,55 @@ func createKeyword(content string) *html.Node {
+		Attr:     []html.Attribute{},
+	}
+	if class != "" {
+		// img.Attr = append(img.Attr, html.Attribute{Key: "class", Val: class})
```
Left over from old code, fixed thx

mrsdizzie commented:
> custom emoji not renderd on review comment:
> 
> ![Screenshot at 2020-04-11 03-04-24](https://user-images.githubusercontent.com/24977596/79031736-3bb6ac80-7ba1-11ea-9e7d-8a06add2174b.png)

Hmm we must not run those through the markdown editor -- will see about that

6543 commented:
> ...What's your distro?

@silverwind artix

silverwind commented:
@6543 Maybe [this](https://dev.to/darksmile92/get-emojis-working-on-arch-linux-with-noto-fonts-emoji-2a9) will help you 😉

6543 commented:
> @6543 Maybe [this](https://dev.to/darksmile92/get-emojis-working-on-arch-linux-with-noto-fonts-emoji-2a9) will help you 😉

Thanks, I already did this when i noticed i need a new font for that.

But my point is: we need a fallback - there are more people out in tje world witch will have a similar expirience and in general most user's with tiss issue will mage gitea responsible for that

silverwind commented:
I don't see fallbacks as strictly neccessary nowadays, but they may be nice to have. I wonder if GitHub's fallback PNGs are open sourced somewhere.

mrsdizzie commented on custom/conf/app.ini.sample:
```diff
@@ -172,8 +172,8 @@ SHOW_USER_EMAIL = true
 THEMES = gitea,arc-green
-; All available reactions. Allow users react with different emoji's
-; For the whole list look at https://gitea.com/gitea/gitea.com/issues/8
+;Reactions to comments. Values can be emoji alias (:smile:) or a unicode emoji.
```
sure fixed

silverwind commented on custom/conf/app.ini.sample:
```diff
@@ -172,7 +172,8 @@ SHOW_USER_EMAIL = true
 ; All available themes. Allow users select personalized themes regardless of the value of `DEFAULT_THEME`.
 THEMES = gitea,arc-green
-;Reactions to comments. Values can be emoji alias (:smile:) or a unicode emoji.
+;All available reactions users can choose on issues/prs and comments.
```
```suggestion
;List of available reactions users can choose on issues/prs and comments.
```

silverwind commented on custom/conf/app.ini.sample:
```diff
@@ -172,7 +172,8 @@ SHOW_USER_EMAIL = true
 DEFAULT_THEME = gitea
 ; All available themes. Allow users select personalized themes regardless of the value of `DEFAULT_THEME`.
```
```suggestion
; List of available themes. Allow users select personalized themes regardless of the value of `DEFAULT_THEME`.
```

silverwind commented on docs/content/doc/advanced/config-cheat-sheet.en-us.md:
```diff
@@ -128,7 +128,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `THEMES`:  **gitea,arc-green**: All available themes. Allow users select personalized themes
   regardless of the value of `DEFAULT_THEME`.
-- `REACTIONS`: All available reactions users can choose on issues and comments
+- `REACTIONS`: All available reactions users can choose on issues/prs and comments
```
```suggestion
- `REACTIONS`: List of available reactions users can choose on issues/prs and comments
```

silverwind commented on docs/content/doc/advanced/config-cheat-sheet.en-us.md:
```diff
@@ -128,7 +128,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `DEFAULT_THEME`: **gitea**: \[gitea, arc-green\]: Set the default theme for the Gitea install.
 - `THEMES`:  **gitea,arc-green**: All available themes. Allow users select personalized themes
```
```suggestion
- `THEMES`:  **gitea,arc-green**: List of available themes. Allow users select personalized themes
```

mrsdizzie commented:
Id be more interested in a proper web font before using images. You'd never replace any other valid unicode with png files and it makes everything unnecessarily more complicated.

We already do this several other places:
 
 https://github.com/go-gitea/gitea/blob/6cf71ee0e8178a2775d2d1041912ca45372805fb/web_src/less/_base.less#L2

silverwind commented:
Yeah, font seems the cleaner method but I'd like it to avoid the font download when one of the system emoji fonts is present, so I guess that would need a little JS to detect the installed fonts and dynamically add a fallback font to the font stack. We don't have to do this in this PR.

mrsdizzie commented:
> Yeah, font seems the cleaner method but I'd like it to avoid the font download when one of the system emoji fonts is present, so I guess that would need a little JS to detect the installed fonts and dynamically add a fallback font to the font stack. We don't have to do this in this PR.

Can't this be done in css/less and don't we already do it in the example above?

```
local('Roboto'),
        local('Roboto-Regular'),
        /* Chrome 26+, Opera 23+, Firefox 39+ */
        url('../vendor/assets/roboto-fonts/roboto-v20-latin-ext_cyrillic-ext_latin_greek_vietnamese_cyrillic_greek-ext-regular.woff2') format('woff2'),
        /* Chrome 6+, Firefox 3.6+, IE 9+, Safari 5.1+ */
        url('../vendor/assets/roboto-fonts/roboto-v20-latin-ext_cyrillic-ext_latin_greek_vietnamese_cyrillic_greek-ext-regular.woff') format('woff');
}
```
Which says download it if you don't already have a local copy? Maybe I misunderstand how that all is working

silverwind commented:
Fallback via CSS may potentially work, we have to test. Question is if something like this would incur a font download on systems where `"Apple Color Emoji"` is present:

```css
body {
  font-family: "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
}

@font-face {
  font-family: "Noto Color Emoji";
  src: local('Noto Color Emoji'), local('Noto-Color-Emoji'), url('...');
}
```

Noto color variant is 9MB, so I'd definitely want to avoid its download if possible:

https://github.com/googlefonts/noto-emoji/blob/master/fonts/NotoColorEmoji.ttf

6543 commented:
I'm looking forward to the webfont

silverwind commented:
@6543 let me know if https://jsfiddle.net/silverwind/n6rdvcoa/ works for your case with the font absent. It does work for me but only in Chrome. Firefox rejects the font:

````
downloadable font: no supported glyph shapes table(s) present (font-family: "Noto Color Emoji" style:normal weight:400 stretch:100 src index:2) source: https://unpkg.com/noto-color-emoji@1.0.1/ttf/NotoColorEmoji.ttf
downloadable font: rejected by sanitizer (font-family: "Noto Color Emoji" style:normal weight:400 stretch:100 src index:2) source: https://unpkg.com/noto-color-emoji@1.0.1/ttf/NotoColorEmoji.ttf
````

Note that the example does not use the current version of the font, which could be a reason it fails in Firefox.

6543 commented:
https://jsfiddle.net/silverwind/n6rdvcoa/ works like a charm @silverwind :tada:

6543 commented:
`:tada:` != :tada:

silverwind commented:
Some things I've noticed using `🐫 :camel: :gitea:` test string:

1. Issue title and timeline entries (main page and issue) do not show emojis (it actually converts unicode back to token?)

<img width="434" alt="image" src="https://user-images.githubusercontent.com/115237/79054189-2db96800-7c43-11ea-8420-c79db00c1eb4.png">
<img width="554" alt="image" src="https://user-images.githubusercontent.com/115237/79054193-37db6680-7c43-11ea-89a2-277051dbbbe3.png">
<img width="164" alt="image" src="https://user-images.githubusercontent.com/115237/79054196-432e9200-7c43-11ea-8c36-2735856d6993.png">

2. inside markdown `:gitea:` (the only custom emoji that should use image) does not work:

<img width="108" alt="image" src="https://user-images.githubusercontent.com/115237/79054210-5f323380-7c43-11ea-9c3e-fe438018365a.png">

In my version I had copied a `emoji.png` to `public/img` to cover that one with an `<img>`.

silverwind commented:
Tribute extraction done. `:gitea:` works in it:

<img width="196" alt="image" src="https://user-images.githubusercontent.com/115237/79054847-8c351500-7c48-11ea-89d2-92c33a7c8de0.png">

I also made +1 and -1 sort first like GitHub:

<img width="205" alt="image" src="https://user-images.githubusercontent.com/115237/79054855-a5d65c80-7c48-11ea-903a-0fce24f8f81d.png">

@mrsdizzie: I would appreciate if you get the JSON format to

````js
{
  "+1": "👍",
  "-1": "👎",
 /* other emojis alphabetically sorted */
}
````
`emoji.js` will need a few tweaks to deal with the new format so it exports `emojiMap` and `emojiKeys` from the JSON.

silverwind commented:
@6543 fallback added, please test.

mrsdizzie commented:
@silverwind why does the JSON need to change? Something to do with the ordering? or is it about which ones show up when you haven't typed anything in?

mrsdizzie commented:
I will look at the added code and other issues when I can

silverwind commented:
> why does the JSON need to change

Primarily for the size reduction but the ordering would be nice to have too because then it's less work on the client side, but not really important.

6543 commented:
@silverwind fallback css is not applyed to code comments

silverwind commented:
@6543 you mean code in general? I don't think we want to do token replacement in code (e.g. all places that use a monospace font) at all, GitHub also does not do that:

````
:tada:
````

6543 commented:
> @6543 you mean code in general? I don't think we want to do token replacement in code (e.g. all places that use a monospace font) at all, GitHub also does not do that:
> 
> ```
> :tada:
> ```

no this is as it should here is the screenshot:
![Screenshot at 2020-04-12 16-11-40](https://user-images.githubusercontent.com/24977596/79070809-53dd1780-7cd8-11ea-833a-9bf75d35503d.png)

mrsdizzie commented:
> @silverwind fallback css is not applyed to code comments

This ishttps://github.com/go-gitea/gitea/issues/10213 

Code review has no love as you can see in screenshot. I'll see if I can fix that as part of this PR

mrsdizzie commented:
> Some things I've noticed using `🐫 :camel: :gitea:` test string:
> 
> 1. Issue title and timeline entries (main page and issue) do not show emojis (it actually converts unicode back to token?)

Hmm Issues titles aren't really meant to be post processed and change text, and I'm not really sure its expected to use any type of short code for emojis  in them (normal emojis work fine). I'll think on it but my first thought is that the shortcode isn't necessarily meant to work anywhere you can type text and I don't think titles should be post processed like that.

I'm not sure what part converts unicode back to a token, but the issue title itself doesn't do that for me either on issue page or timeline (using the 🐫 :camel: :gitea: and others).

If theres an example of a specific timeline event that does it that can be fixed (if it is a review comment thats a noted bug below). I suspect there are lots of random things that aren't run through a markdown processor and none of this code will run for those.

guillep2k commented:
1.11.4 does support short codes in PR/issue titles (used `test :camel:` as text).

![image](https://user-images.githubusercontent.com/18600385/79091476-ca533700-7d23-11ea-81c3-a73c3e1245bb.png)

```
gitea=# select * from issue where id  = 179;
 id  | repo_id | index | poster_id |     name     |      content      | milestone_id | priority | is_closed | is_pull | num_comments | ref | deadline_unix | created_unix | updated_unix | closed_unix | is_locked | original_author | original_author_id
-----+---------+-------+-----------+--------------+-------------------+--------------+----------+-----------+---------+--------------+-----+---------------+--------------+--------------+-------------+-----------+-----------------+--------------------
 179 |     109 |     2 |         9 | Test :camel: | Test with :camel: |            0 |        0 | f         | f       |            0 |     |             0 |   1586750828 |   1586750828 |           0 | f         |                 |                  0
(1 row)
```

mrsdizzie commented:
> > why does the JSON need to change
> 
> Primarily for the size reduction but the ordering would be nice to have too because then it's less work on the client side, but not really important.

Sure, I've sorted them which will put it in that order. I still don't want to change the JSON from how it is to save the >10KB and want to keep the aliases as mentioned above.

For moving to assets -- it seems everything else is in public and I assumed that was because files public can be hosted on another server via setting. Is there a reason this wouldn't apply for the json file? I only see one file in assets currently

mrsdizzie commented:
> 1.11.4 does support short codes in PR/issue titles (used `test :camel:` as text).
> 
> ![image](https://user-images.githubusercontent.com/18600385/79091476-ca533700-7d23-11ea-81c3-a73c3e1245bb.png)

It works now because the javascript just replaces most anything on the page with little context of what it is.

I guess I'm wondering should it work going forward, because we've (and every other site) has already made a decision to intentionally not post process titles and change their text/contents from what a user enters otherwise.

But it could be added if that is really desired, I've no strong opinion other than that.

mrsdizzie commented on web_src/js/features/emoji.js:
```diff
@@ -0,0 +1,23 @@
+export const giteaImage = `<img alt=":gitea:" title=":gitea:" class="emoji" src="${StaticUrlPrefix}/img/emoji.png" align="absmiddle">`;
+
+const tempMap = {gitea: 'gitea'};
+for (const {emoji, aliases} of emojis) {
```
I still want to have access to the description field as before I think it makes a better text result in the menu but more importantly it makes a better aria-label and I want to use it for that

silverwind replied:
I specifically changed it to use the keyword because that matches GitHub's behaviour. I think the descriptions are not really useful and would mostly just be confusing. Everyone seems to refer to emojis based on the description.

I think it may be nice to have `title="keyword"` tooltip, that's something I've wanted on GitHub too.

mrsdizzie replied:
Sorry, I like the description better for at least the aria label here and do want to have access to it here as it was before.

mrsdizzie commented:
@silverwind other than comment about description these changes look good thanks much for taking the time to make them

silverwind commented:
> it seems everything else is in public and I assumed that was because files public can be hosted on another server via setting. Is there a reason this wouldn't apply for the json file

There's no real reason to have it in `public` (which would also duplicate it into `bindata` needlessly) because it's imported and compiled into the JS bundle using the `'../../../public/emoji/emoji.json'` import, so it can be anywhere in the repo.

mrsdizzie commented:
> > it seems everything else is in public and I assumed that was because files public can be hosted on another server via setting. Is there a reason this wouldn't apply for the json file
> 
> There's no real reason to have it in `public` (which would also duplicate it into `bindata` needlessly) because it's imported and compiled into the JS bundle using the `'../../../public/emoji/emoji.json'` import, so it can be anywhere in the repo.

Ah OK I see I will move into assets then

silverwind commented:
So the question is really how to do the replacement in general. Things we want to support emoji:

- Any place where markdown is supported
- Issue titles
- All places where the above to appear rendered

Things we do not want to support replacement:

- Rendered Code (file view, inline blocks in markdown)

There was https://github.com/go-gitea/gitea/issues/10601 which requests to support a subset of markdown in issue titles, so I guess we could have issue titles have backtick and emoji support, nothing else. I guess that can be done by hand without a markdown processor involved.

silverwind commented on web_src/js/features/tribute.js:
```diff
@@ -0,0 +1,61 @@
+      return `:${item.original}:`;
+    },
+    menuItemTemplate(item) {
+      return `<div class="tribute-item"><span class="emoji">${item.original === 'gitea' ? giteaImage : emojiMap[item.original]}</span>${item.original}</div>`;
```
FYI I plan to refactor this into a shared function that returns emoji HTML from a keyword.

silverwind replied:
And it's done.

mrsdizzie commented:
You can apply any of the individual processors if configured, so it would be easy to add just the emoji ones for a title I'll do that

silverwind commented:
Commit title may be another canidate for replacement, I remember that working on GH too.

mrsdizzie commented:
> Commit title may be another canidate for replacement, I remember that working on GH too.

I've added it to all post processing functions so it already works here and anywhere else that is processed (Issue titles just weren't before). 

I've moved emoji.json to assets and added initial support for issue titles. It works in main places but I'll need to go through and manually update timeline events, cross references, and wherever else its shows the title

silverwind commented:
Something gone wrong with my latest change as it includes other commits, will fix with a rebase against master.

silverwind commented:
One fix pushed for issue title where the `1.5em` proved too much:

Before:
![image](https://user-images.githubusercontent.com/115237/79125930-3de65a00-7d9f-11ea-97ea-d5897816cd25.png)

After:
![image](https://user-images.githubusercontent.com/115237/79125947-45a5fe80-7d9f-11ea-8a74-650561962480.png)

mrsdizzie commented on web_src/js/features/emoji.js:
```diff
@@ -21,3 +19,16 @@ export const emojiMap = {};
+export function emojiHTML(name) {
+  let inner;
+  if (name === 'gitea') {
+    inner = `<img class="emoji" alt=":${name}:" src="${StaticUrlPrefix}/img/emoji.png" align="absmiddle">`;
```
why is this emoji.png and not gitea.png?

silverwind replied:
gitea.png could be customized by the user for the site-wide icon and I think we don't want to render the custom image in this case.

mrsdizzie replied:
I guess I wonder why anybody would change gitea.png to be something other than the provided gitea logo or if they did do that for some reason it is probably then what they would want to see here as well.

:gitea: is going to render to gitea.png in the code always so it would be weird to show a different image in the tribute in the off case somebody did make a change to it.

silverwind replied:
Thought I may undo this and just use `gitea.png` as I think that's how it previously worked (using potentially customized image variant).

mrsdizzie commented:
> One fix pushed for issue title where the `1.5em` proved too much:

thanks that looks great! Will probably have to do the same for events on the timeline when they get added as well since those emoji shouldn't be made larger either

silverwind commented:
> Will probably have to do the same for events on the timeline

Maybe it'd wise to only size-increase `.markdown .emoji` by `1.5em` and default the others to `1em`, but I guess we may need to increase others a bit more in some cases.

silverwind commented:
Still need to sync the server- and client-side rendering (server has `aria-label` while client now has `title` attribute). I'd still favor a reduced version of `emoji.json` and not use the description at all (only keywords).

silverwind commented:
Another place for replacement is page `<title>`:

![](https://user-images.githubusercontent.com/115237/79126996-49d31b80-7da1-11ea-9d50-34fd47da6aa0.png)

silverwind commented:
To highlight the size issue: `index.js` goes from 452kB to 585kB with this change. Yes, it might be much less of a difference when gzipped but it's still something the JS parsers has to parse so it will add bit of a delay to the JS initialization.

guillep2k commented:
> Things we **do not want** to support replacement:

IMHO: commit messages

6543 commented on templates/repo/issue/view_content/add_reaction.tmpl:
```diff
@@ -4,17 +4,11 @@
-			{{else}}
-				<div class="item" data-content="{{$value}}">:{{$value}}:</div>
-			{{end}}
+				<div class="item reaction" data-content="{{$value}}">{{ReactionToEmoji $value}}</div>
```
```suggestion
			<div class="item reaction" data-content="{{$value}}">{{ReactionToEmoji $value}}</div>
```

6543 commented:
Crome Console: `Tribute was already bound to TEXTAREA`

6543 commented:
and code-review comments html are not paresd for emoji

guillep2k commented:
>
> I've moved emoji.json to assets and added initial support for issue titles. It works in main places but I'll need to go through and manually update timeline events, cross references, and wherever else its shows the title

Mail templates... 🙄

silverwind commented:
Rebased to resolve conflict with master.

silverwind commented:
> Crome Console: Tribute was already bound to TEXTAREA

IIRC, that error is also present on master.

lunny commented:
> To highlight the size issue: `index.js` goes from 452kB to 585kB with this change. Yes, it might be much less of a difference when gzipped but it's still something the JS parsers has to parse so it will add bit of a delay to the JS initialization.

How about generate a new file `emoji.js`?

silverwind commented:
> How about generate a new file emoji.js?

It could potentially be coupled to `RequireSimpleMDE` and lazy-loaded but I'd actually prefer to load it unconditionally but in the small variant (30kB).

mrsdizzie commented on web_src/js/features/tribute.js:
```diff
@@ -0,0 +1,61 @@
+    },
+    selectTemplate(item) {
+      if (typeof item === 'undefined') return null;
+      return `:${item.original}:`;
```
@silverwind sorry not sure when this was added but this should insert the actual emoji character as it did before and not the short code. Lots of these changes are breaking or changing what was existing intentional and desired behavior which wasn't really my intention when asking for help making the javascript better and not what I would like for this PR

silverwind replied:
Actually, I did not touch this one yet but you're right it should, except for the `:gitea:` case which will only show when the markdown is rendered.

silverwind replied:
done in https://github.com/go-gitea/gitea/pull/11032/commits/5f30a0277e40d4a998e4189c611a717465c99a82.

silverwind replied:
Also somewhat related: https://github.com/go-gitea/gitea/issues/11081

mrsdizzie replied:
Thanks!

and interesting for that ticket, perhaps the short code converter can see what is in the config file and not do anything if it isn't utf8mb4 (though I assume these people are just getting general errors whenever they have users try and use an emoji or import some content that has them)

silverwind replied:
Yeah it is already affecting existing data that has emojis, it's just that they will be more prominent now with the token conversion in place.

silverwind commented:
It appears the fallback causes Firefox to spew the error mentioned in https://github.com/go-gitea/gitea/pull/11032#issuecomment-612485673 into the console, even when better fonts are present. I'll probably attempt to load the fallback via JS to circumvent this (thought it is essentially a browser bug).

silverwind commented:
Actually, the fallback download is only triggered for symbols where the native font does not have it. For example [accordion](https://emojipedia.org/emojipedia/13.0/accordion/) added in Unicode 13 which I currently do not have on my platform. [Fallback example](https://jsfiddle.net/silverwind/n6rdvcoa/) updated with this.

I'm thinking we maybe want to limit the available emoji by maximum unicode version using a constant in the code, for example I can not select `accordion` on GitHub either, so they must be doing that too.

mrsdizzie commented:
> Actually, the fallback download is only triggered for symbols where the native font does not have it. For example [accordion](https://emojipedia.org/emojipedia/13.0/accordion/) added in Unicode 13 which I currently do not have on my platform. [Fallback example](https://jsfiddle.net/silverwind/n6rdvcoa/) updated with this.
> 
> I'm thinking we maybe want to limit the available emoji by maximum unicode version using a constant in the code, for example I can not select `accordion` on GitHub either, so they must be doing that too.

We removed the unicode version for size ; ) I can add that field back and I assume the detection would be done in the Javascript and just not show that option in the tribute menu?

silverwind commented:
I was thinking you could define `MAX_UNICODE_VERSION` and just omit any characters higher than that variable?

mrsdizzie commented:
> `MAX_UNICODE_VERSION`

Wouldn't you have the same problem if I set a global variable to 13? I'd think the only way you could deal with this is storing the unicode version for each emoji as Github does and not including them in the results if a user doesn't have that version.

silverwind commented:
GitHub currently supports Unicode 12 and my browser only supports 11, so I don't think they're doing any kind of detection and it's really just a build-time constant.

I'm pretty sure unicode version detection in JS is [hard if not impossible](https://stackoverflow.com/questions/1911000/detecting-individual-unicode-character-support-with-javascript).

I guess it's not a big deal if you want to ship with Unicode 13, I just wanted to avoid that fallback console error.

mrsdizzie commented:
Ah I see ok that makes sense -- I'll set it at 12 and use that field to filter them out of the final data we use then

mrsdizzie commented:
@silverwind ok now only including up to unicode 12 and I give you a smaller JSON too 🤏

mrsdizzie commented:
@6543 fixed all template/dashboard issues you mentioned, reviews too : ) lemme know if you see more

@guillep2k mail templates too 📫

6543 commented:
one problem remain, a bug with +1,-1
see at discord chat since github dont store png currently ?!?

6543 commented:
same bug with titles

mrsdizzie commented:
@6543 sorry was typo, fixed 👍

6543 commented:
![Screenshot at 2020-04-16 04-43-45](https://user-images.githubusercontent.com/24977596/79409005-0ba24b80-7f9d-11ea-8a5e-1197e2b204bd.png)

6543 commented:
I think you changed things in the right place :) now only thumbsdown has an issue

lafriks commented on build/generate-emoji.go:
```diff
@@ -34,9 +35,19 @@ type Gemoji []Emoji
+	UnicodeVersion string   `json:"unicode_version,omitempty"`
+}
+
+// Don't unclude some fields in JSON
```
```suggestion
// Don't include some fields in JSON
```

lafriks commented on modules/templates/helper.go:
```diff
@@ -85,6 +86,21 @@ func NewFuncMap() []template.FuncMap {
 		"AllowedReactions": func() []string {
 			return setting.UI.Reactions
 		},
+		"ReactionToEmoji": func(reaction string) template.HTML {
```
This should also probably be moved to function like `RenderEmoji`

mrsdizzie replied:
They are slightly different -- ReactionToEmoji just returns just the emoji character for use in the existing reactions template and RenderEmoji returns different html with full <span class=emoji> etc...

Some of these seem similar but there are slightly different use cases for them

silverwind replied:
I also have two distinct functions in JS, one that returns HTML and one that returns a unicode string. Maybe rename the helpers for clarity, `RenderEmojiString`, `RenderEmojiHTML`.

mrsdizzie replied:
One converts a 'reaction' as defined in the "REACTIONS" config setting to an emoji character without any span tag etc... and the other renders a full span tag/string for use in almost every other place in templates. Both those functions return template.HTML and not strings. 

There is a third one that returns a plain string and it is RenderEmojiPlain which is needed for page title and email subject that needs to just be a string.

They all use the same underlying functions from emoji.go just return slightly different things as needed.

lafriks replied:
I did not mean that they are the same, I meant that you move this function below and leave only function name here

mrsdizzie replied:
Ah OK I see what you mean sorry I was confused. Fixed : )

silverwind commented:
Will review again later.

mrsdizzie commented:
Theres one small issue I'll try and fix later where when writing to json file zero width joiners work properly but when writing to go file they are not combined properly in some cases. The only thing this actually effects is the description not being included as an attr on a subset of otherwise detected and rendered emoji but I want to fix it so I can make the tests better.

mrsdizzie commented:
@lafriks ok this is now done as far as I am concerned :)

Learned more than I wanted to know about unicode and various invisible characters but tests are now complete and also eliminate any false positive

lafriks commented:
@mrsdizzie see my comment above

lafriks commented on modules/templates/helper.go:
```diff
@@ -85,6 +86,9 @@ func NewFuncMap() []template.FuncMap {
 		"AllowedReactions": func() []string {
 			return setting.UI.Reactions
 		},
+		"RenderEmojiPlain": func(s string) string {
```
```suggestion
		"RenderEmojiPlain": emoji.ReplaceAliases
```

mrsdizzie replied:
fixed

silverwind commented:
This is almost LGTM, but `:gitea:` does not seem to work at all for me:

<img width="263" alt="Screen Shot 2020-04-17 at 21 54 34" src="https://user-images.githubusercontent.com/115237/79608953-263c0800-80f6-11ea-863f-9f5a69311bbf.png">

silverwind commented:
@GiteaBot is dumb, don't land yet.

mrsdizzie commented:
@silverwind I changed the location of the image, do you perhaps have some sort of stale cache thing happening? works fine here

silverwind commented:
@mrsdizzie I don't think so, it does not seem to attempt replacement at all.

mrsdizzie commented:
> @mrsdizzie I don't think so, it does not seem to attempt replacement at all.

oh you know what it might be, its never actually done it unless you have gitea in your reactions settings. Is it that? I guess I never planned on having it hardcoded into the go code but since it is hardcoded into js I guess it could be there too

silverwind commented:
Yeah my reaction settings are default and I just tested on master and `:gitea:` does work with defaults.

mrsdizzie commented:
> Yeah my reaction settings are default and I just tested on master and `:gitea:` does work with defaults.

Should be fixed

mrsdizzie commented:
For some reason I don't understand this is making a few integration tests on arm64 take longer to the point the overall testing times out. There aren't any errors from the tests themselves so not sure what is happening : (

lafriks commented:
I think when multiple drone jobs are run in parallel it seems gets slow imho, idk to be honest

mrsdizzie commented:
@lafriks it is something else, you can see some tests like ```TestAPIDeleteFile```taking a very long time for some reason but they don't fail. The don't have trouble on amd64 either.

Not sure how that could be related to this PR though but something seems to make it take longer

silverwind commented:
Rebased and fixed one merge conflict.

silverwind commented:
@mrsdizzie would you mind editing the issue references in the first message to match the commit description, just so GitHub correctly associates the issues?

````
Fixes: https://github.com/go-gitea/gitea/issues/9182
Fixes: https://github.com/go-gitea/gitea/issues/8974
Fixes: https://github.com/go-gitea/gitea/issues/8953
Fixes: https://github.com/go-gitea/gitea/issues/6628
Fixes: https://github.com/go-gitea/gitea/issues/5130
````

guillep2k commented:
Last couple of times I've relaunched the CI it was stuck on `TestGit/SSH/BranchProtectMerge` (PostgreSQL and SQLite3 only, in the AMD container). Definitively reproducible. No idea about the cause. Maybe we should check the stack traces for clues.

mrsdizzie commented:
@guillep2k its not really stuck on those it is just the tests get killed after 40 minutes thats where it is at then -- some of the tests are running much longer (as noted above) for some reason that would seem unrelated to this PR but is certainly reproducible every time -- but also didn't exist at the start of this PR either (and there has been no large code change since then mostly small fixes).

Wonder if there is some similarity to https://github.com/go-gitea/gitea/pull/10906 where that made arm build fail then take 10 minutes longer for no reason then just seemingly went away with no change to the PR (interestingly too this started happening after that PR was merged but not before, though it doesn't effect other PRs).

Think we would need the Gitea logs from that drone container/run if possible to see why some of those tests are taking so long and where they are spending their time waiting -- the tests themselves don't fail so there is no helpful log otherwise.

guillep2k commented:
@mrsdizzie Yeah, the logs from Drone don't contain timestamps, so it's difficult to detect what's taking so long. Within this subject, I think we should disable certain submodules while testing, and enable them only for the specific tests: repo indexing, issue indexing, scheduled tasks (e.g. cron), etc. I suspect we'll gain some speed there.

guillep2k commented:
Could you edit `.drone.yml` and increase the timeouts for pgsql/sqlite, just for the kicks? e.g.:

```
- timeout -s ABRT 50m make test-pgsql-migration test-pgsql
```

techknowlogick commented:
@guillep2k Drone does contain timestamps, but you’ll need to go into “full screen” to see the time elapsed in the log view on the right hand side. Lmk if you need more details on how to access that info.

mrsdizzie commented:
@guillep2k it isn't obvious at first but there are timestamps in drone:

 * visit: https://drone.gitea.io/go-gitea/gitea/24028
 * click sqlite in arm64 section
 * click the play button in the top right corner
 * then make full screen

You can see when each test starts so for example you can see that ```TestAPIDeleteFile``` starts at 316s and then next test doesn't start until 536s -- implying (I think??) that ```TestAPIDeleteFile```takes 220 seconds to finish. This happens on a few tests to the point that after 40 minutes we kill it. I'm not really sure how this PR would effect such tests (or why it wouldn't happen for the amd64 tests).

The only thing I could think to do is see the actual gitea logs to see what part of those tests is taking so long and go from there (I don't have an arm64 to test on and can't reproduce this otherwise)

guillep2k commented:
@techknowlogick @mrsdizzie Thanks! I've missed it because I was expecting something like date+time, rather than elapsed. 👍

guillep2k commented:
SQLite:
![image](https://user-images.githubusercontent.com/18600385/79644217-e1f44a80-817d-11ea-804c-3afcc25f8cbe.png)

MySQL:
![image](https://user-images.githubusercontent.com/18600385/79644223-e6206800-817d-11ea-8773-8aa21efa0f65.png)

Any  ideas?

mrsdizzie commented:
Would there be some sort of disk write slowing down/limiting this PR is hitting on arm64 (maybe because it already modifies so many files before starting?)

It seems to be those tests and also ```TestGit/HTTP/BranchProtectMerge``` that all take around 220-260S to run and finish

mrsdizzie commented:
@guillep2k @lafriks we can probably stop merging master into this PR for now until underlying write limit issues are fixed on ARM64. 

I don't want to take up other users CI time when this seems guaranteed to fail with current limits. When those issues are fixed I'll resolve any conflicts if needed for this PR.

As for any further debugging, my guess is because this PR deletes a thousand files real quick (removing all emojify img files) that it makes the CI run be identified as some sort of high disk write process and then forces delays in writing to help balance things out -- which is why it doesn't happen for other PR. Just a guess though nothing else to base that on

lafriks commented:
Maybe we could just merge it without drone check

mrsdizzie commented:
We could try : ) if the problem is something else it might cause trouble for other CI runs after that though

lafriks commented:
@lunny @techknowlogick what do you think?

6543 commented:
I'll `make test-sqlite` localy

techknowlogick commented:
I would rather wait until CI passes to merge for reason @mrsdizzie said, in case it is some other issue. I think what we could do is to run the build via drone cli on an arm64 server. One of the @go-gitea/owners could do this (or maybe provide temporary access to @go-gitea/mergers to the arm64 server), or if someone else wants to dig into this I have coupons for $100 of packet.com spend so they could spin up their own test just lmk (just note the servers are expensive so make sure you only provision it when you are testing as that credit wouldn't last a full month)

6543 commented:
Local test result: **PASS**
```
real	4m32,361s
user	4m43,215s
sys	0m54,789s
```
with tmpfs

guillep2k commented:
I agree with @mrsdizzie. This problem will most likely be dragged along with any future PRs.

I wonder if there's some tweaking we can do, like with the npx stuff (looks expensive), or maybe remove some tests in certain configurations (e.g. raw git stuff should be pretty much engine agnostic, it shouldn't require testing in all engines).

mrsdizzie commented:
This disk limiting issue seems to still happen even when reverting the removal of many files from this PR so perhaps it is something else triggering disk issues. I spun up an arm64 instance at packet but the test runs fine (much faster even) there on its own:

```
[test-sqlite:361] PASS
[test-pgsql:375] PASS
```

I don't think I can easily replicate the same type of I/O as our CI running many PRs at once to see if it picks this one to slow down.  I think without specific logs from current build server as to why it doesn't like this PR I am out of ideas : (

zeripath commented:
@mrsdizzie perhaps it may be worth squashing/force-pushing?

lafriks commented:
@zeripath that won't change anything as drone anyway checks out only latest branch version

6543 commented:
can we just increase the timeout on drone?

mrsdizzie commented:
@6543 that won't help -- we already know the tests pass and it isn't a problem with that. 

The problem seemingly that is something about this PR is causing disk writes to become very delayed by the system which makes any test that writes files take a very long time to complete. The worry is that if you merge this PR the problem could continue and then become a problem in every CI run because the code is now part of all builds. Either every build would timeout and fail or they would all take 60+ minutes.

That might not happen and the problem could just be part of the PR process itself, but because it isn't clear what the specific problem is that would be the risk in merging it.

silverwind commented:
I don't quite buy this "file io" theory. When we install node_modules, we install around 50k files and I don't think it has a significant effect on the overall runtime.

But you could try changing [this](https://github.com/go-gitea/gitea/blob/7179e498815bcfc0eaa81dc08c499be6ddc3dc2f/.drone.yml#L304) to just `make backend` and see if it helps any. Frontend is already fully tested in `compliance` so it does not need to build again during the integration tests.

silverwind commented:
Removed a few useless frontend build invokations. Let's see how https://drone.gitea.io/go-gitea/gitea/24186 goes.

mrsdizzie commented:
Seems it would have to be something unique about this PR since it only happens here so I wouldn't imagine it is related to any front end building that happens on all other PRs.

The tests that are drastically slowed down compared to other runs don't call any code from this PR at all, all involve writing files, and appear (best guess anyway) to be a result of some sort of write limiting/delayed writing according to server stats which show i/o capacity at 100%. 

You can see delays start with the first test to write files (TestAPICreatePullSuccess1) and just get worse from there on future tests that write files. Only happens here though so must be something about it CI doesn't like 🧐

silverwind commented:
Do we set [`DRONE_CPU_QUOTA`](https://docs.drone.io/runner/docker/configuration/reference/drone-cpu-quota/)? Is there anything in this PR that could cause excessive CPU usage maybe?

mrsdizzie commented:
Added some basic debugging to at least try and show its disk writes being drastically slowed down from some sort of throttling and not directly waiting or stuck on any code from this PR:

 https://drone.gitea.io/go-gitea/gitea/24239/3/5 -> testing-arm64->sqlite

```
=== TestAPIDeleteFile (integrations/api_repo_file_delete_test.go:40)
Test deleting file and making the delete in a new branch
165		Creating file [delete/file3.txt]
166		Time taken: [11.884673386s]
167		Deleting file [delete/file3.txt]
168		Time taken: [22.286854178s]
169	Time taken: [34.20145366s]
```
etc.. Those are all from modules->repofiles->update.go:CreateOrUpdateRepoFile but it just seems to be writes in general (other tests that write are slow too I just didn't add any debugging output)
 
And:
```
=== TestUndoDeleteBranch (integrations/branches_test.go:35)
381	in emoji processor
382	Time taken: [17.88µs]
383	in emoji processor
384	in emoji processor
385	Time taken: [8.08µs]
386	in emoji processor
387	in emoji processor
388	Time taken: [9.46µs]
```
Perhaps if done or something higher than the system level does have throttling built in there would be a log of it being activated somewhere?

mrsdizzie commented:
Further debugging suggest the specific command responsible for 99% of this time spent in these slow writes:

```
Calling RunInDirTimeoutEnvFullPipelineFunc

 command: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= fetch origin --no-tags master:base master:original_base 

Time taken: [8.429074179s]
```

Creating and cloning a temporary repo do not seem to suffer any:
```
Creating temporary upload repository

  Cloning temporary upload repository

      Calling RunInDirTimeoutEnvFullPipelineFunc

      command: /usr/bin/git -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= -c credential.helper= -c protocol.version=2 -c filter.lfs.required= -c filter.lfs.smudge= -c filter.lfs.clean= clone -s --bare -b master /tmp/tmp.dyuGXWVMtj/integrations/gitea-integration-sqlite/gitea-repositories/user2/repo1.git integrations/gitea-integration-sqlite/data/tmp/local-repo-sqlite/upload.git339962698

      Time taken: [16.736059ms]

...cut...

  Time taken: [17.01576ms]

Time taken: [4.021084ms] -- total time to create temp repo
```
Creating temporary repos doesn't seem to show any slowness

guillep2k commented:
Notice that changes we do in the drone file are not used by the CI while testing this PR. AFAIK the pipeline goes like this:

- Start from whatever is in `master`.
- Run an instance with that.
- First step: merge this PR into the _working copy_.

So, unless drone re-reads the `.yml` config, I suspect none of our changes will have an effect until merged. 😬

mrsdizzie commented:
@guillep2k you can see from this PR it uses any changes to drone.yml when doing CI (including not running if you have a config error 😬)

CI for this PR is currently only running one specific integration test just for debugging since this problem only happens in our CI and not anywhere else : ( 

But I think that is probably as far as I could get from this end to show general disk throttling/delay is happening and not directly from any code in this PR. Something on the server would need to be changed or made clear else I think it will be stuck here.

guillep2k commented:
@mrsdizzie It seems that you are right; Drone appears to re-read the `.yml` file after the merge (check that checking out the PR is the first step).

As for your test, it only failed because the mysql step `integration-test-coverage` is not prepared for running individual tests with `#test` notation.

I consider that your hypothesis about the I/O cap is correct, or at least closely related to the actual problem.

guillep2k commented on .drone.yml:
```diff
@@ -183,7 +181,7 @@ steps:
       - "curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | bash"
       - apt-get install -y git-lfs
-      - make test-mysql-migration integration-test-coverage
+      - make test-mysql-migration integration-test-coverage#TestAPIDeleteFile
```
```suggestion
      - make test-mysql-migration integration-test-coverage
```

Can you remove this here to see it succeeding? I'm marking it as a change request to prevent accidental merging.

mrsdizzie replied:
There is no concern for mysql tests for this particular problem it won't show anything new and having it run all of the tests would just take away more unnecessary amd64 CI time while debugging.

The only thing that matters for now is how long it takes to write files on arm64 which is still way off unfortunately -- even just running just the one test it already appears to have hit some kind of limit or other issue that makes writing very slow.

zeripath commented:
@mrsdizzie is the push back triggering hooks?

Is it possible that the integration tests are causing a reinitialisation of Gitea or are just timing out?

mrsdizzie commented:
@zeripath I don't think so the tests PASS and API sends back expected response so nothing is timing out/dying. Just taking 15-30s to do each one

silverwind commented:
Maybe worth checking if changing `resources` in `drone.yml` has any effect. It's very hard to find any documentation on Drone's `resources` or if they even work on the Docker runner. Some things I've found from the Kubernetes docs:

https://docs.drone.io/pipeline/kubernetes/syntax/steps/#resources.
https://docs.drone.io/runner/docker/configuration/reference/drone-cpu-quota/
https://www.kernel.org/doc/Documentation/scheduler/sched-bwc.txt

According to CFS docs:

> cpu.cfs_quota_us: the total available run-time within a period (in microseconds)
> cpu.cfs_period_us: the length of a period (in microseconds)

So assuming `cfs_period_us` is left the default `100000`, I think this may try, which should limit the process to four CPU cores:

````yaml
steps:
- name: build
  image: golang
  resources:
    limits:
      cpu: 400000
````

cc: @tboerger

mrsdizzie commented:
Added that but it doesn't seem to help -- but at this point all this PR is doing is building Gitea and running a single integration test which doesn't use any of the code from the PR.

It also doesn't seem to effect any other Drone build either on amd64 during our CI or when I tried testing on a another one of these arm64 servers at packet.com (and nobody can reproduce it at home).

silverwind commented:
Another idea might be to try with a Alpine image like `image: golang:1.14-alpine` just to see if musl libc makes any difference.

silverwind commented on .eslintrc:
```diff
@@ -23,6 +23,7 @@ globals:
   emojify: false
```
```suggestion
```

Missed that one

mrsdizzie commented:
One more round of debugging shows git itself seems to be waiting random amounts of time between running hooks on push. As an example, here is the first test from TestAPICreateFile (full debug output in drone result if interested). For the time consuming parts:

Normal run from amd64:
```
13:12:33.072213 run-command.c:643       trace: run_command: GIT_ALTERNATE_OBJECT_DIRECTORIES=/tmp/tmp.cHkWvZNcmk/integrations/gitea-integration-mysql/gitea-repositories/user2/repo1.git/./objects GIT_OBJECT_DIRECTORY=/tmp/tmp.cHkWvZNcmk/integrations/gitea-integration-mysql/gitea-repositories/user2/repo1.git/./objects/incoming-ZSSPoC GIT_PUSH_OPTION_COUNT=0 GIT_QUARANTINE_PATH=/tmp/tmp.cHkWvZNcmk/integrations/gitea-integration-mysql/gitea-repositories/user2/repo1.git/./objects/incoming-ZSSPoC hooks/pre-receive
	remote: 	time taken: 3.303679ms        
	remote: 
13:12:33.133928 run-command.c:643       trace: run_command: hooks/update refs/heads/master 65f1bf27bc3bf70f64657658635e66094edbcb4d 57bafb39742effb4e08a865723696cbde77e0d9d
	remote: 13:12:33.194513	 running hookUpdate        
	remote: 	time taken: 29.591µs        
	remote: 
	13:12:33.198339 run-command.c:643       trace: run_command: GIT_PUSH_OPTION_COUNT=0 hooks/post-receive
	remote: 13:12:33.255899 running hookPostReceive        
	remote: 2020-04-24 13:12:33.257670 running private HookPostReceive        
	remote: 13:12:33.257689 making internal request:        
	remote: 13:12:33.378192 request finished:        
	remote: time taken: 120.619454ms        
	remote: 
	remote: hook post-receive time taken: 122.400904ms        
	remote: 
	remote: . Processing 1 references        
	remote: Processed 1 references in total        
13:12:33.381923 run-command.c:643       trace: run_command: git gc --auto --quiet
```

Bad slow run from arm64:
```
13:13:09.090739 run-command.c:643       trace: run_command: GIT_ALTERNATE_OBJECT_DIRECTORIES=/tmp/tmp.XCTSiGWl0x/integrations/gitea-integration-sqlite/gitea-repositories/user2/repo1.git/./objects GIT_OBJECT_DIRECTORY=/tmp/tmp.XCTSiGWl0x/integrations/gitea-integration-sqlite/gitea-repositories/user2/repo1.git/./objects/incoming-orubLg GIT_PUSH_OPTION_COUNT=0 GIT_QUARANTINE_PATH=/tmp/tmp.XCTSiGWl0x/integrations/gitea-integration-sqlite/gitea-repositories/user2/repo1.git/./objects/incoming-orubLg hooks/pre-receive
	remote: 	time taken: 10.989025ms        
	remote: 
13:13:12.590502 run-command.c:643       trace: run_command: hooks/update refs/heads/master 65f1bf27bc3bf70f64657658635e66094edbcb4d 57bafb39742effb4e08a865723696cbde77e0d9d
	remote: 13:13:15.943662	 running hookUpdate        
	remote: 	time taken: 107.661µs        
	remote: 
13:13:15.960425 run-command.c:643       trace: run_command: GIT_PUSH_OPTION_COUNT=0 hooks/post-receive
	remote: 13:13:22.073344 running hookPostReceive        
	remote: 2020-04-24 13:13:22.078827 running private HookPostReceive        
	remote: 13:13:22.078848 making internal request:        
	remote: 13:13:22.279336 request finished:        
	remote: time taken: 200.763179ms        
	remote: 
	remote: hook post-receive time taken: 206.273412ms        
	remote: 
	remote: . Processing 1 references        
	remote: Processed 1 references in total        
13:13:22.297067 run-command.c:643       trace: run_command: git gc --auto --quiet
```

So its taking about 13S to do the same thing as it takes 300ms in a normal run. The times it takes Gitea to run the hooks doesn't change though.

Notice several second delays between running each hook, despite Gitea itself returning in just ms. Even the update hook which does nothing at all but return nil in gitea still takes 2 seconds before git starts the next one. The CI is full of debugging information like this which shows various delays between running hooks, sometimes longer sometimes faster. 

Doesn't appear to be Gitea stuck on anything and git trace logging doesn't suggest it is doing anything between the time Gitea returns from a hook and the time git runs the next one. Still a mystery 🕵️‍♀️

guillep2k commented:
Ping LG-TM!
